Lutron component overhaul and improvements

cdheiser opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
- Upgrade pylutron to 0.2.7
- Add support for configuring via the integrations UI
- Bring in new unique IDs from pylutron
- Properly define devices vs. entities
  - Keypads, Switches, Main Repeater, and Remotes are all devices
- Add support for importing Area names and assignments from Lutron
  - Users can enable this to import areas and then turn it off to
    override specific devices.
- Add services for press-and-hold, release, and tapping of buttons
  - This allows for the use of the alarm button where holding the
    holding the button is required to keep the alarm state active.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/13962

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

cdheiser committed:
Add a new service for pressing Lutron Buttons and generate unique IDs.

1) Buttons
The LutronButton has been updated to support press (and hold), release,
and tap events.  This allows automations and other integrations to
activate buttons like an alarm.

2) Unique IDs.
pylutron 0.2.6 added support for unique IDs.  These have been added
to the entities here.

cdheiser committed:
Improve Area Support

- Make it possible to ask the AreaRegistry to return an area or create it if one does not exist.
- Make it possible for a device to assert which area it should be in.
  This is useful for integration with platforms that also have a concept of areas.

cdheiser committed:
Lutron component overhaul

- Upgrade pylutron to 0.2.6
- Add support for configuring via the integrations UI
- Bring in new unique IDs from pylutron
- Properly define devices vs. entities
  - Keypads, Switches, Main Repeater, and Remotes are all devices
- Add support for importing Area names and assignments from Lutron
  - Users can enable this to import areas and then turn it off to
    override specific devices.
- Add services for press-and-hold, release, and tapping of buttons
  - This allows for the use of the alarm button where holding the
    holding the button is required to keep the alarm state active.

JonGilmore commented:
Excellent @cdheiser! I'll do some testing of this later today and get back to you

cdheiser commented:
I've rebased and fixed some of the lint errors.  I have some outstanding PRs for pylutron to get in to clean up the  rest.

cdheiser committed:
Properly update-after-add for lutron devices.

I accidentally dropped the update-after-add for adding entities with the
overhaul to use config entries and async code.

JonGilmore commented:
Just getting around to testing this out. Things are looking good, but I'm seeing some issues with text displaying correctly. For example, when you click options on the Lutron integration, you'll see empty text here: 
![image](https://user-images.githubusercontent.com/7232986/85235095-ef28f200-b3d7-11ea-8e51-097310517584.png)

Or, when you add the integration via the UI, it's missing text description labels as well: 
![image](https://user-images.githubusercontent.com/7232986/85235126-24354480-b3d8-11ea-83ce-ca7ee43cf016.png)

cdheiser committed:
Only add a binary_sensor to areas that have sensors

The new version of pylutron properly binds an OccupancyGroup to all areas as defined in the XML.

cdheiser committed:
Don't include the area name in devices when we're letting lutron define and assign devices to areas.

cdheiser commented:
So that's because translations are missing, and it's not clear to me if I'm
supposed to check in the ones that the local dev tools produce, or if that
should get handled by another process (it's not really well documented)

On Sun, Jun 21, 2020 at 1:59 PM Jon Gilmore <notifications@github.com>
wrote:

> Just getting around to testing this out. Things are looking good, but I'm
> seeing some issues with text displaying correctly. For example, when you
> click options on the Lutron integration, you'll see empty text here:
> [image: image]
> <https://user-images.githubusercontent.com/7232986/85235095-ef28f200-b3d7-11ea-8e51-097310517584.png>
>
> Or, when you add the integration via the UI, it's missing text description
> labels as well:
> [image: image]
> <https://user-images.githubusercontent.com/7232986/85235126-24354480-b3d8-11ea-83ce-ca7ee43cf016.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/36902#issuecomment-647180872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWXN5UIQ5EENDQVEMFLRXZYCZANCNFSM4OBIDMRQ>
> .
>

shenxn commented:
@cdheiser You can use
```
python3 -m script.translations develop
```
to generate translation files to test locally. But do not commit the generated files.

cdheiser commented:
That's what I thought.  Thanks for confirming.  So @JonGilmore you can generate the right translation files locally to fill in the words.

And you really need to try turning on the option to use areas from Lutron.  I can't believe it took me this long to suggest this PR.

cdheiser committed:
Update to pylutron 0.2.7 which has fixes for UUIDs for OccupancyGroups

cdheiser committed:
Fix the TODO to get the name of the controller via a getter instead of accessing the protected class member.

cdheiser commented:
This should now be ready to go.  pylutron was updated today as was this PR

JonGilmore commented:
thanks @cdheiser - i'll be testing this out today and I'll get back to you with any feedback

cdheiser commented:
What's next?

cdheiser commented:
@homeassistant What's next?

cdheiser commented:
@JonGilmore you have any idea how to get someone assigned to accept and merge this PR?

JonGilmore commented:
@MartinHjelmare can you help us in getting this reviewed?

cdheiser commented:
Ping?

On Sat, Aug 1, 2020, 6:10 AM Jon Gilmore <notifications@github.com> wrote:

> @MartinHjelmare <https://github.com/MartinHjelmare> can you help us in
> getting this merged?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/36902#issuecomment-667529811>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWUMIRXYKV7RGKH6MXLR6QH53ANCNFSM4OBIDMRQ>
> .
>

JonGilmore commented:
@frenck could you help us out?

cdheiser commented:
Ping?  Do I need to file a bug to get someone to review this PR?

On Sat, Aug 8, 2020, 9:50 AM Jon Gilmore <notifications@github.com> wrote:

> *@JonGilmore* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/36902#pullrequestreview-463814268>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWVEKL3KYHSQ64VT4CLR7V637ANCNFSM4OBIDMRQ>
> .
>

MartinHjelmare commented:
Please be patient. We'll get to it.

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -4,13 +4,14 @@
+from homeassistant import config_entries
 from homeassistant.const import ATTR_ID, CONF_HOST, CONF_PASSWORD, CONF_USERNAME
-from homeassistant.helpers import discovery
+from homeassistant.helpers import area_registry as ar, device_registry as dr
```
Please use longer variable names than two characters.

joshtbernstein replied:
@MartinHjelmare I'm asking to learn, but this [example](https://developers.home-assistant.io/docs/device_registry_index/#manual-registration) shows something very similar, so is the example incorrect?

MartinHjelmare replied:
This particular case is fine. Normally we want at minimum three characters.

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
+
+async def async_setup_entry(hass, entry):
+    """Set up a single Lutron deployment."""
+    if not entry.data:
```
Why would entry.data be falsy? That can't happen.

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
+    config = entry.data
+    device_registry = await dr.async_get_registry(hass)
+    area_registry = await ar.async_get_registry(hass)
+    hass_data = {
```
Please name this something that's not so easily confused with `hass.data`.
```suggestion
    lutron_data = {
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
+        },
+    }
+    hass.data[DOMAIN][entry.entry_id] = hass_data
+    hass_data[LUTRON_CONTROLLER] = Lutron(
```
```suggestion
    controller = hass_data[LUTRON_CONTROLLER] = Lutron(
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
 
-    hass.data[LUTRON_CONTROLLER].load_xml_db()
-    hass.data[LUTRON_CONTROLLER].connect()
+    await hass.async_add_executor_job(hass_data[LUTRON_CONTROLLER].load_xml_db)
```
```suggestion
    await hass.async_add_executor_job(controller.load_xml_db)
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
-    hass.data[LUTRON_CONTROLLER].load_xml_db()
-    hass.data[LUTRON_CONTROLLER].connect()
+    await hass.async_add_executor_job(hass_data[LUTRON_CONTROLLER].load_xml_db)
+    hass_data[LUTRON_CONTROLLER].connect()
```
Is this async safe?
```suggestion
    controller.connect()
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
 
     # Sort our devices into types
-    for area in hass.data[LUTRON_CONTROLLER].areas:
+    for area in hass_data[LUTRON_CONTROLLER].areas:
```
```suggestion
    for area in controller.areas:
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -36,36 +37,74 @@
+        if entry.options.get(CONF_ENABLE_AREAS, DEFAULT_ENABLE_AREAS):
+            hass_area = area_registry.async_get_or_create(area.name)
+        else:
+            hass_area = DummyArea(area.name)
```
This seems weird. Creating areas and mixing them with dummy areas. I'm not sure we allow setting areas at all from integrations.

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
 
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        # connections={(dr.CONNECTION_NETWORK_MAC, config.mac)},
```
Commented code.
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+        identifiers={(DOMAIN, hass_data[LUTRON_CONTROLLER].guid)},
+        manufacturer="Lutron",
+        name=hass_data[LUTRON_CONTROLLER].name,
+        # model=config.modelid,
+        # sw_version=config.swversion,
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        # connections={(dr.CONNECTION_NETWORK_MAC, config.mac)},
+        identifiers={(DOMAIN, hass_data[LUTRON_CONTROLLER].guid)},
+        manufacturer="Lutron",
+        name=hass_data[LUTRON_CONTROLLER].name,
```
```suggestion
        identifiers={(DOMAIN, controller.guid)},
        manufacturer="Lutron",
        name=controller.name,
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+        # sw_version=config.swversion,
+    )
+
+    _LOGGER.debug("Loading Components")
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+    )
+
+    _LOGGER.debug("Loading Components")
     for component in ("light", "cover", "switch", "scene", "binary_sensor"):
```
Please store these names in a constant list.

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+    _LOGGER.debug("Loading Components")
     for component in ("light", "cover", "switch", "scene", "binary_sensor"):
-        discovery.load_platform(hass, component, DOMAIN, {}, base_config)
+        _LOGGER.debug("Loading %s", component)
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    _LOGGER.debug("Setup Services!!!")
+    setup_services(hass, hass_data)
+    _LOGGER.debug("Setup Entry returning true...")
```
```suggestion
    async_setup_services(hass, hass_data)
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -80,31 +119,81 @@ def setup(hass, base_config):
+ATTR_NAME = "button"
+
+
+def setup_services(hass, hass_data):
```
```suggestion
def async_setup_services(hass, hass_data):
```

MartinHjelmare commented on homeassistant/components/lutron/__init__.py:
```diff
@@ -119,23 +208,53 @@ def _update_callback(self, _device, _context, _event, _params):
+            },
+            "name": self.name,
+            "manufacturer": "Lutron",
+            # "model": self.light.productname,
+            # "sw_version": self.light.swversion,
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+
+        return self.async_create_entry(title=user_input[CONF_HOST], data=user_input)
+
+    def _show_setup_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
```
Please inline the form in the user step.

MartinHjelmare commented on homeassistant/components/lutron/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+        if user_input is None:
+            return self._show_setup_form()
+
+        await self.async_set_unique_id(user_input[CONF_HOST])
```
Ip address is not allowed as unique id. Is this an ip address?

MartinHjelmare replied:
If there's no other good unique id, we can just check the existing config entries and make sure we don't set up another entry with the same host. Ie do this without using the unique id api.

MartinHjelmare commented on homeassistant/components/lutron/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+        await self.async_set_unique_id(user_input[CONF_HOST])
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(title=user_input[CONF_HOST], data=user_input)
```
Before creating the entry we need to validate the connection to the controller.

MartinHjelmare commented on homeassistant/components/lutron/scene.py:
```diff
@@ -38,4 +46,25 @@ def activate(self, **kwargs: Any) -> None:
+            "name": f"{self._area.name} {self._keypad.name}",
+            "manufacturer": "Lutron",
+            "model": self._keypad.type,
+            # "sw_version": self.light.swversion,
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/lutron/switch.py:
```diff
@@ -107,3 +119,22 @@ def update(self):
+            "name": f"{self._area.name} {self._keypad.name}",
+            "manufacturer": "Lutron",
+            "model": self._keypad.type,
+            # "sw_version": self.light.swversion,
```
```suggestion
```

MartinHjelmare commented on homeassistant/helpers/area_registry.py:
```diff
@@ -48,6 +48,18 @@ def async_list_areas(self) -> Iterable[AreaEntry]:
         return self.areas.values()
 
+    @callback
+    def async_get_or_create(self, name: str) -> AreaEntry:
```
This needs to be approved first by the core team. Has this been discussed?

cdheiser replied:
> Have the area registry changes been discussed with the core team? I'm not sure we'll approve those.

They have not.  How would you suggest we proceed on that specifically?  (I could separate them out into a different PR).  This is incredibly useful for those of us with systems that have their own notion of areas.  Generally speaking, this was the entire motivation for this entire overhaul.

I'll go over the other requested changes and config_flow tests.

MartinHjelmare replied:
Please open an architecture issue.
https://github.com/home-assistant/architecture/issues

MartinHjelmare replied:
Sounds good to remove it from this PR if possible so we aren't blocked by that review.

cdheiser replied:
I've created https://github.com/home-assistant/architecture/issues/425 and a PR specifically for that functionality.

I'll rebase this PR and then split out the area integration support and work on the comments.

MartinHjelmare reviewed (changes requested):
Tests for the config flow are missing. Please use the scaffold script to get started with that.

Have the area registry changes been discussed with the core team? I'm not sure we'll approve those.

cdheiser commented:
I plan to get back to revising this in the next week or so.

chrisaljoudi commented:
@cdheiser @MartinHjelmare I know & appreciate you guys are busy—would you mind quickly sharing what all still needs to be done here? Happy to try and help.

MartinHjelmare commented:
The comments above need to be addressed.

balloob commented:
This PR seems to have gone stale. Closing it.

joshtbernstein commented:
@cdheiser There's some good changes here. Would you have any interest in merging them before the area stuff goes through? I'm happy to help in anyway that I'm able to.

cdheiser commented:
I do have interest in getting changes merged in, I'm just short on time at
the moment.

On Wed, Feb 10, 2021 at 2:26 PM Josh <notifications@github.com> wrote:

> @cdheiser <https://github.com/cdheiser> There's some good changes here.
> Would you have any interest in merging them in a new PR before the area
> stuff goes through? I'm happy to help in anyway that I'm able to.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/36902#issuecomment-777077352>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWXFZQATIM66XI67GOLS6MB2JANCNFSM4OBIDMRQ>
> .
>

joshtbernstein commented:
What would be the proper way to pick this up? Do I fork HA and manually add in your changes or do I start with your branch?

cdheiser commented:
I'm attempting to rebase this PR.  It's been a while and thus I may not be
successful, but if I can rebase it successfully, it'll be an easier place
to start from.

On Mon, Mar 15, 2021 at 3:30 PM Josh ***@***.***> wrote:

> What would be the proper way to pick this up? Do I fork HA and manually
> add in your changes or do I start with your branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/36902#issuecomment-799797929>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQARWV7BO67Q42H6QBRQB3TD2C7VANCNFSM4OBIDMRQ>
> .
>

cdheiser commented:
I believe https://github.com/cdheiser/home-assistant/tree/pylutron-upgrade
should now be both caught up to reality, and should work properly with
suggested_area support that was added.  (I haven't tested it at all yet).

On Sat, Mar 20, 2021 at 10:18 PM Chris Heiser ***@***.***>
wrote:

> I'm attempting to rebase this PR.  It's been a while and thus I may not be
> successful, but if I can rebase it successfully, it'll be an easier place
> to start from.
>
> On Mon, Mar 15, 2021 at 3:30 PM Josh ***@***.***> wrote:
>
>> What would be the proper way to pick this up? Do I fork HA and manually
>> add in your changes or do I start with your branch?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/home-assistant/core/pull/36902#issuecomment-799797929>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACQARWV7BO67Q42H6QBRQB3TD2C7VANCNFSM4OBIDMRQ>
>> .
>>
>

joshtbernstein commented:
@cdheiser 

You have a typo [here](https://github.com/cdheiser/home-assistant/blob/ee4b3f3fa412631f714713b435c7cb8111944034/homeassistant/components/lutron/__init__.py#L209), but after correcting that it loads and adds devices. I'll do some more testing over the next few days. Should we be tracking this somewhere else?
